### ISSUE TITLE
Do not use babel-preset-react-optimize with react-hot-loader

### ISSUE THREAD

**Issue by Kannnnng** [gaearon/react-hot-loader#890]
Do not use `babel-preset-react-optimize` with `react-hot-loader`, it will cause that React's Component lose their state and replace the whole root element of React.

I don't know why. This is the result of my experiment.

**Comment by theKashey** [comment]
This might be due to `transform-react-constant-elements`. it just "freezes" components, moving all the "static" stuff off the render, thus not letting RHL do it's work.

**Comment by Schlesiger** [comment]
I tried _individually_ using the `babel` maintained plugins from that preset to optimize my bundles. I was testing with the `^7.0.0-beta.0` releases.

I was able to use `@babel/plugin-transform-react-constant-elements` in both development and production without apparent impact to `react-hot-loader`. It didn't "freeze" or "reload" my state on component updates. However, it is recommended in the [documentation](https://babeljs.io/docs/plugins/transform-react-constant-elements/) to only deploy in production due to making "warning messages more cryptic".

I did have trouble with `@babel/plugin-transform-react-inline-elements` breaking `react-hot-loader` in development _even_ if it was set to only run in production. If I would run a production build beforehand, `react-hot-loader` would fail in my development build. However, if I deleted the cache before starting my development build, everything would work.

The plugin also seems to break component libraries. If I shared a production build of a component built with `@babel/plugin-transform-react-inline-elements`, it would break `react-hot-loader` for those downstream users.

That being said, the console errors were about components unrelated to my project. If I tried a simple demo with no component dependencies; everything worked properly. I started bringing in the components that were listed in the errors into the small demo project; trying to make it break. I couldn't. I'm not sure at what combination of `@babel/plugin-transform-react-constant-elements` and the component tree breaks `react-hot-loader`.

That is why I'm not 100% confident in these findings, but I know if I don't use `@babel/plugin-transform-react-inline-elements`, everything works normally.
